src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILInterpreter.cs

```diff
@@ -2126,5 +2118,337 @@ private void InterpretSwitch(ref ILReader reader, int jmpBase, int[] jmpDelta)
                     reader.Seek(jmpBase + jmpDelta[i]);
             }
         }
+
+        private void InterpretNewArray(int token)
+        {
+            int length = 0;
+            StackItem stackItem = PopWithValidation();
+
+            switch (stackItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    length = stackItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    length = (int)stackItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            Debug.Assert(length >= 0);
+
+            TypeDesc arrayType = ((TypeDesc)_methodIL.GetObject(token)).MakeArrayType();
+
+            // TODO: Add support for arbitrary  non-primitive types
+            Array array = RuntimeAugments.NewArray(arrayType.GetRuntimeTypeHandle(), length);
+
+            _stack.Push(StackItem.FromObjectRef(array));
+        }
+
+        private void InterpretLoadLength()
+        {
+            Array array = (Array)PopWithValidation().AsObjectRef();
+            _stack.Push(StackItem.FromInt32(array.Length));
+        }
+
+        private void InterpretStoreElement(ILOpcode opcode)
+        {
+            StackItem valueItem = PopWithValidation();
+            StackItem indexItem = PopWithValidation();
+            Array array = (Array)PopWithValidation().AsObjectRef();
+
+            int index = 0;
+            switch (indexItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    index = indexItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    {
+                        long value = (long)indexItem.AsNativeInt();
+                        Debug.Assert(value >= int.MinValue && value <= int.MaxValue);
+                        index = (int)value;
+                    }
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            if (index < 0 || index >= array.Length)
```

Comment:
Nit: This can be `if ((uint)index >= (uint)array.Length)` to make it a bit faster. We use this trick frequently for array bound checks.

Reply:
Maybe this check can be inside `GetSzArrayElementAddress`.

Reply:
> We use this trick frequently for array bound checks.

Oh I saw it on Dictionary, is there also some special jit treatment if it recognize pattern?

Reply:
It is one comparison instead of two. It would be nice if the JIT recognized the two comparisons and converted them into one, but we are not there yet.